common/src/main/java/bisq/common/crypto/KeyStorage.java

```diff
@@ -118,9 +135,51 @@ public KeyPair loadKeyPair(KeyEntry keyEntry) {
                 //noinspection ResultOfMethodCallIgnored
                 fis.read(encodedPrivateKey);
 
+                // Read magic bytes
+                byte[] magicBytes = Arrays.copyOfRange(encodedPrivateKey, 0, ENCRYPTED_FORMAT_MAGIC.length);
+                boolean isEncryptedPassword = Arrays.compare(magicBytes, ENCRYPTED_FORMAT_MAGIC) == 0;
+                if (isEncryptedPassword && password == null) {
+                    throw new IncorrectPasswordException("Cannot load encrypted keys, user must open account with password " + filePrivateKey);
+                } else if (password != null && !isEncryptedPassword) {
+                    log.warn("Password not needed for unencrypted key " + filePrivateKey);
+                }
+
+                // Decrypt using password
+                if (password != null) {
+                    int position = ENCRYPTED_FORMAT_MAGIC.length;
+
+                    // Read remaining header
+                    ByteBuffer buf = ByteBuffer.wrap(encodedPrivateKey, position, ENCRYPTED_FORMAT_LENGTH);
+                    position += ENCRYPTED_FORMAT_LENGTH;
+                    int version = buf.getInt();
+                    if (version != 1) throw new RuntimeException("Unable to parse encrypted keys");
+                    int saltLength = buf.getInt();
+                    int iterations = buf.getInt();
+                    int secretLength = buf.getInt();
+
+                    byte[] salt = Arrays.copyOfRange(encodedPrivateKey, position, position + saltLength);
+                    position += saltLength;
+                    KeyCrypterScrypt crypter = ScryptUtil.getKeyCrypterScrypt(salt);
+                    KeyParameter pwKey = ScryptUtil.deriveKeyWithScrypt(crypter, password);
+                    byte[] pwEncrypted = Arrays.copyOfRange(encodedPrivateKey, position, position + pwKey.getKey().length);
+                    if (Arrays.compare(pwEncrypted, pwKey.getKey()) != 0) {
+                        throw new IncorrectPasswordException("Incorrect password");
+                    }
+                    position += pwEncrypted.length;
+
+                    // Payload salt
+                    salt = Arrays.copyOfRange(encodedPrivateKey, position, position + saltLength);
+                    position += saltLength;
+
+                    // Decrypt payload
+                    SecretKey secretKey = Encryption.generateSecretKey(password, salt, iterations, secretLength);
+                    byte[] encryptedPayload = Arrays.copyOfRange(encodedPrivateKey, position, encodedPrivateKey.length);
+                    encodedPrivateKey = Encryption.decryptPayloadWithHmac(encryptedPayload, secretKey);
```

Comment:
`decryptPayloadWithHmac` should detect if the password is incorrect. Is the operation done with ScryptUtil necessary?

Reply:
see `throw new CryptoException("Hmac does not match.");` unfortunately you'd probably have to check the message of the exception is "Hmac does not match" to see it's the password and not something else?

Reply:
It's been awhile so I will need to double check, but from my memory I recalled the error / exception resulting from an incorrect password decryption was not reliable or accurate as that was how I first approached it. I will give it another try and follow up.

Reply:
It looks like if I attempt to decrypt using an incorrect password, I get this exception:

`ERROR bisq.common.crypto.KeyStorage: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption. 
bisq.common.crypto.CryptoException: javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
	at bisq.common.crypto.Encryption.decrypt(Encryption.java:98)
	at bisq.common.crypto.Encryption.decryptPayloadWithHmac(Encryption.java:173)
	at bisq.common.crypto.KeyStorage.loadKeyPair(KeyStorage.java:178)
	at bisq.common.crypto.KeyRing.unlockKeys(KeyRing.java:96)
	at bisq.core.api.CoreAccountService.openAccount(CoreAccountService.java:100)
	at bisq.daemon.app.HavenoDaemonMain.interactiveLogin(HavenoDaemonMain.java:193)`

If we assume the exception means the key was generated with a bad password, I could handle the exception and remove the password hash. In addition I could use Scrypt key derivation to do the encrypting. Would simplify things a bit.

Reply:
I've updated the code to use the scrypt drived key to do the encryption along with the hmac decryption. It is no longer necessary.